src/pymor/reductors/loewner.py

```diff
@@ -0,0 +1,255 @@
+
+    r_cc = np.array([], dtype=int)
+    for ri in right:
+        if np.imag(s[ri]) != 0:
+            r_cc = np.concatenate((r_cc, np.where(s == s[ri].conj())[0]))
+    right = np.concatenate((right, r_cc))
+
+    return (left, right)
+
+
+def loewner_quadruple(s, Hs, partitioning='even-odd', ordering='regular', ltd=None, rtd=None):
+    r"""Constructs a Loewner quadruple as |NumPy arrays|.
+
+    The Loewner quadruple :cite:`ALI17`
+
+    .. math::
+        (\mathbb{L},\mathbb{L}_s,V,W)
+
+    consists of the Loewner matrix :math:`\mathbb{L}`, the shifted Loewner matrix
+    :math:`\mathbb{L}_s`, left interpolation data :math:`V` and right interpolation data :math:`W`.
+
+    Parameters
+    ----------
+    s
+        |Numpy Array| of shape (n,) containing the frequencies.
+    Hs
+        |Numpy Array| of shape (n, p, m) or |TransferFunction| resembling the transfer function
+        samples.
+    partitioning
+        `str` or `tuple` of length 2. Strings can either be 'even-odd' or 'half-half' defining
+        the splitting rule. A user-defined partitioning can be defined by passing a tuple of the
+        left and right indices. Defaults to `even-odd`.
+    ordering
+        The ordering with respect to which the splitting rule is executed. Can be either
+        'magnitude', 'random' or 'regular'. Defaults to 'regular'.
+    ltd
+        |Numpy Array| representing left tangential directions. If `None` tangential directions will
+        be chosen randomly (normal distribution).
+    rtd
+        |Numpy Array| representing right tangential directions. If `None` tangential directions will
+        be chosen randomly (normal distribution).
+
+    Returns
+    -------
+    L
+        Loewner matrix.
+    Ls
+        Shifted Loewner matrix.
+    V
+        Left interpolation data.
+    W
+        Right interpolation data.
+    """
+    assert isinstance(s, np.ndarray)
+    if hasattr(Hs, 'transfer_function'):
+        Hs = Hs.transfer_function
+    assert isinstance(Hs, TransferFunction) or isinstance(Hs, np.ndarray) and Hs.shape[0] == s.shape[0]
+
+    assert partitioning in ('even-odd', 'half-half') or len(partitioning) == 2
+    assert ordering in ('magnitude', 'random', 'regular')
+
+    if isinstance(Hs, TransferFunction):
+        Hss = np.empty((len(s), Hs.dim_output, Hs.dim_input), dtype=s[0].dtype)
+        for i, ss in enumerate(s):
+            Hss[i] = Hs.eval_tf(ss)
+        Hs = Hss
+    else:
+        assert len(Hs[0]) == len(s)
+
+    # ensure that complex sampling values appear in complex conjugate pairs
+    for i, ss in enumerate(s):
+        if np.conj(ss) not in s:
+            s = np.append(s, np.conj(ss))
+            Hs = np.append(Hs, np.conj(Hs[i])[None, ...], axis=0)
+
+    ip, jp = _partition_frequencies(s, Hs, partitioning, ordering) if isinstance(partitioning, str) else partitioning
+
+    if len(Hs.shape) > 1:
+        dim_output = Hs.shape[1]
+        dim_input = Hs.shape[2]
+        if ltd is None:
```

Comment:
It would be good to enable non-tangential MIMO Loewner.

Reply:
Agreed, it should be part of this PR.